drivers/usb/device/usb_dc_usbip.c

```diff
@@ -0,0 +1,922 @@
+static int usbip_handle_devlist(int connfd)
+{
+	struct usb_desc_header *desc_hdr = __usb_descriptor_start;
+	struct usb_if_descriptor *if_desc = NULL;
+	struct usbip_devlist_iface_data iface;
+	struct usbip_devlist_header rep_hdr = {
+		.version = htons(USBIP_VERSION),
+		.code = htons(USBIP_OP_REP_DEVLIST),
+		.status = 0,
+		.ndev = htonl(1),
+	};
+	int rc;
+
+	LOG_DBG("Handle OP_REQ_DEVLIST");
+
+	rc = send(connfd, &rep_hdr, sizeof(rep_hdr), 0);
+	if (rc != sizeof(rep_hdr)) {
+		return -errno;
+	}
+
+	rc = send(connfd, &devlist_data, sizeof(devlist_data), 0);
+	if (rc != sizeof(devlist_data)) {
+		return -errno;
+	}
+
+	while (desc_hdr->bLength != 0U) {
+		if (desc_hdr->bDescriptorType == USB_INTERFACE_DESC) {
+			if_desc = (struct usb_if_descriptor *)desc_hdr;
+
+			iface.bInterfaceClass = if_desc->bInterfaceClass;
+			iface.bInterfaceSubClass = if_desc->bInterfaceSubClass;
+			iface.bInterfaceProtocol = if_desc->bInterfaceProtocol;
+			iface.padding = 0U;
+
+			rc = send(connfd, &iface, sizeof(iface), 0);
+			if (rc != sizeof(iface)) {
+				return -errno;
+			}
+		}
+
+		/* Move to next descriptor */
+		desc_hdr = (struct usb_desc_header *)((uint8_t *)desc_hdr +
+						      desc_hdr->bLength);
+	}
+
+	return 0;
+}
+
+static int usbip_handle_import(int connfd)
+{
+	struct usbip_req_header rep_hdr = {
+		.version = htons(USBIP_VERSION),
+		.code = htons(USBIP_OP_REP_IMPORT),
+		.status = 0,
+	};
+	char busid[32];
+	int rc;
+
+	LOG_DBG("Handle OP_REQ_IMPORT");
+
+	rc = recv(connfd, &busid, sizeof(busid), MSG_WAITALL);
+	if (rc <= 0) {
+		return rc == 0 ? -ENOTCONN : -errno;
+	}
+
+	LOG_HEXDUMP_DBG(busid, sizeof(busid), "busid");
+
+	rc = send(connfd, &rep_hdr, sizeof(rep_hdr), 0);
+	if (rc != sizeof(rep_hdr)) {
+		return -errno;
+	}
+
+	rc = send(connfd, &devlist_data, sizeof(devlist_data), 0);
+	if (rc != sizeof(devlist_data)) {
+		return -errno;
+	}
+
+	usbip_ctrl.cmd_num_in = 0;
+	usbip_ctrl.cmd_num_out = 0;
+
+	return 0;
+}
+
+static void usbip_ntoh_submit(struct usbip_command *cmd)
+{
+	cmd->hdr.command = ntohl(cmd->hdr.command);
+	cmd->hdr.seqnum = ntohl(cmd->hdr.seqnum);
+	cmd->hdr.devid = ntohl(cmd->hdr.devid);
+	cmd->hdr.direction = ntohl(cmd->hdr.direction);
+	cmd->hdr.ep = ntohl(cmd->hdr.ep);
+
+	if (cmd->hdr.command == USBIP_CMD_SUBMIT) {
+		cmd->submit.flags = ntohl(cmd->submit.flags);
+		cmd->submit.length = ntohl(cmd->submit.length);
+		cmd->submit.start_frame = ntohl(cmd->submit.start_frame);
+		cmd->submit.numof_iso_pkts = ntohl(cmd->submit.numof_iso_pkts);
+		cmd->submit.interval = ntohl(cmd->submit.interval);
+	} else {
+		cmd->unlink.seqnum = ntohl(cmd->unlink.seqnum);
+	}
+}
+
+static int usbip_handle_submit(int connfd, struct usbip_command *cmd)
+{
+	struct usbip_cmd_submit *req = &cmd->submit;
+	int rc;
+
+	rc = recv(connfd, req, sizeof(*req), MSG_WAITALL);
+	if (rc <= 0) {
+		return rc == 0 ? -ENOTCONN : -errno;
+	}
+
+	usbip_ntoh_submit(cmd);
+	usbip_ctrl.devid = cmd->hdr.devid;
+	usbip_ctrl.connfd = cmd->hdr.seqnum;
```

Comment:
There's a mechanism I don't understand... What is the relevance between `seqnum` and socket FD? Why do we set `seqnum` to the value of socket FD in `usbip_send_common()`? And most importantly, why do we overwrite the socket FD we use for communication with the value of `seqnum`, where the `seqnum` is a value read from a message sent by a peer? It seems like a serious security issue to me.

Reply:
This lines are wrong and was an intermediate state that I pushed too early, I was too careless in vim.